subsys/event_manager/event_manager_priv.h

```diff
@@ -227,6 +221,75 @@ struct event_header {
 	const struct event_type *type_id;
 };
 
+/** Function to log data from this event. */
+typedef void (*log_event_data)(const struct event_header *eh);
+
+/** Deprecated function to log data from this event. */
+typedef	int (*log_event_data_dep)(const struct event_header *eh, char *buf, size_t buf_len);
+
+/** @brief Event type.
+ */
+
+struct event_type {
+	/** Event name. */
+	const char			*name;
+
+	/** Pointer to the array of subscribers. */
+	const struct event_subscriber	*subs_start;
+
+	/** Pointer to the element directly after the array of subscribers. */
+	const struct event_subscriber	*subs_stop;
+
+	/** Bool indicating if the event is logged by default. */
+	bool init_log_enable;
+
+	/** Function to log data from this event. */
+	log_event_data log_event_func;
+
+	/** Deprecated function to log data from this event. */
+	log_event_data_dep log_event_func_dep;
+
+	/** Custom data related to tracking. */
+	const void *trace_data;
+};
+
+extern struct event_type _event_type_list_start[];
+extern struct event_type _event_type_list_end[];
+
+#define _EVENT_TYPE_DEFINE(ename, init_log_en, log_fn, trace_data_pointer)			\
+	_EVENT_SUBSCRIBERS_ARRAY_TAGS(ename);							\
+	STRUCT_SECTION_ITERABLE(event_type, _CONCAT(__event_type_, ename)) = {			\
+		.name               = STRINGIFY(ename),						\
+		.subs_start         = _EVENT_SUBSCRIBERS_START_TAG(ename),			\
+		.subs_stop          = _EVENT_SUBSCRIBERS_END_TAG(ename),			\
+		.init_log_enable    = init_log_en,						\
+		.log_event_func_dep = ((IS_ENABLED(CONFIG_LOG) && Z_C_GENERIC) ?		\
```

Comment:
the disadvantage is that it will use additional memory - it will impact smaller devices